conf/merge.go

```diff
@@ -0,0 +1,180 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package conf
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+
+	"gopkg.in/yaml.v3"
+)
+
+type (
+	// YAML has three fundamental types. When unmarshaled into interface{},
+	// they're represented like this.
+	mapping = map[string]interface{}
+	array   = []interface{}
+	scalar  = interface{}
+)
+
+// mergeYamlFiles deep-merges any number of mergeYamlFiles sources, with later sources taking
+// priority over earlier ones.
+//
+// Maps are deep-merged. For example,
+//
+//	{"one": 1, "two": 2} + {"one": 42, "three": 3}
+//	== {"one": 42, "two": 2, "three": 3}
+//
+// Arrays are appended. For example,
+//
+//	{"foo": [1, 2, 3]} + {"foo": [4, 5, 6]}
+//	== {"foo": [1, 2, 3, 4, 5, 6]}
+func mergeYamlFiles(path string) ([]byte, error) {
+	var merged interface{}
+	var hasContent bool
+
+	apath, err := filepath.Abs(path)
+	if err != nil {
+		return nil, err
+	}
+	files, err := filepath.Glob(apath + "/*.yaml")
+	if err != nil {
+		return nil, err
+	}
+	if len(files) == 0 {
+		return nil, fmt.Errorf("yaml file not found")
+	}
+	for _, f := range files {
+		b, err := ioutil.ReadFile(f)
+		if err != nil {
+			return nil, err
+		}
+		d := yaml.NewDecoder(bytes.NewReader(b))
+
+		var contents interface{}
+		if err := d.Decode(&contents); err == io.EOF {
+			// Skip empty and comment-only sources, which we should handle
+			// differently from explicit nils.
+			continue
+		} else if err != nil {
+			return nil, fmt.Errorf("couldn't decode source: %v", err)
+		}
+
+		hasContent = true
+		pair, err := merge(merged, contents)
+		if err != nil {
+			return nil, err // error is already descriptive enough
+		}
+		merged = pair
+	}
+
+	buf := &bytes.Buffer{}
+	if !hasContent {
+		// No sources had any content. To distinguish this from a source with just
+		// an explicit top-level null, return an empty buffer.
+		return buf.Bytes(), nil
+	}
+	enc := yaml.NewEncoder(buf)
+	if err := enc.Encode(merged); err != nil {
+		return nil, fmt.Errorf("couldn't re-serialize merged YAML: %v", err)
+	}
+	return buf.Bytes(), nil
+}
+
+// merge merges different types using different rules
+func merge(into, from interface{}) (interface{}, error) {
+	// It's possible to handle this with a mass of reflection, but we only need
+	// to merge whole YAML files. Since we're always unmarshaling into
+	// interface{}, we only need to handle a few types. This ends up being
+	// cleaner if we just handle each case explicitly.
+	if into == nil {
+		return from, nil
+	}
+	if from == nil {
+		// Allow higher-priority YAML to explicitly nil out lower-priority entries.
+		return nil, nil
```

Comment:
I don't understand this. it should keep the `into` without change.

Reply:
this is try to rewrite nil in from, but return into is better because merge order is not certain